TiFlash crash in HashJoin

set @@tidb_isolation_read_engines='tiflash,tidb'; set @@tidb_allow_mpp=1;SELECT  table1 . col_varchar_binary_not_null_key  IN ( SELECT 'p' UNION  SELECT 's' ) AS field1 FROM ( C AS table1 RIGHT OUTER JOIN ( ( EE AS table2 INNER JOIN R AS table3 ON (table3 . col_decimal_30_10_not_null = table2 . pk  ) ) ) ON (table3 . col_decimal_not_null_key = table2 . col_int_key  ) ) WHERE (  table3 . col_int_not_null_key >= ( SELECT   SUM(  SUBQUERY2_t1 . col_int ) AS SUBQUERY2_field1 FROM P AS SUBQUERY2_t1  ) ) OR (  ( SELECT   SUM(  SUBQUERY3_t1 . col_varchar_binary_key ) AS SUBQUERY3_field1 FROM ( B AS SUBQUERY3_t1 LEFT  JOIN W AS SUBQUERY3_t2 ON (SUBQUERY3_t2 . col_varchar_binary_key = SUBQUERY3_t1 . col_varchar_binary_not_null  ) ) WHERE ( SUBQUERY3_t1 . col_int_not_null = 1 AND SUBQUERY3_t1 . col_varchar_binary >= SUBQUERY3_t1 . col_varchar_binary_not_null ) ) IS NOT NULL OR table1 . col_varchar_binary_not_null <> table1 . col_varchar_binary_key )   ;

Stack:
[2023-12-31T07:20:28.045Z] [2023/12/31 07:20:11.694 +00:00] [ERROR] [BaseDaemon.cpp:560] ["
       0x77b93b1\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+125539249]
                \tlibs/libdaemon/src/BaseDaemon.cpp:211
  0x7f352e8e0c20\t<unknown symbol> [libpthread.so.0+76832]
       0x7c63090\tDB::Join::handleOtherConditions(DB::Block&, DB::PODArray<unsigned char, 4096ul, Allocator<false>, 15ul, 16ul>*, DB::PODArray<unsigned long, 4096ul, Allocator<false>, 15ul, 16ul>*, std::__1::vector<unsigned long, std::__1::allocator<unsigned long> > const&) const [tiflash+130429072]
                \tdbms/src/Interpreters/Join.cpp:1033
       0x7c809b0\tDB::Join::joinBlock(DB::ProbeProcessInfo&, bool) const [tiflash+130550192]
                \tdbms/src/Interpreters/Join.cpp:2249
       0x88ce3b7\tDB::HashJoinProbeTransformOp::transformHeaderImpl(DB::Block&) [tiflash+143451063]      
                \tdbms/src/Operators/HashJoinProbeTransformOp.cpp:61
       0x899cc2f\tDB::PipelineExecBuilder::appendTransformOp(std::__1::unique_ptr<DB::TransformOp, std::__1::default_delete<DB::TransformOp> >&&) [tiflash+144297007]    
                \tdbms/src/Flash/Pipeline/Exec/PipelineExecBuilder.cpp:28
       0x8a46195\tDB::PhysicalJoinProbe::buildPipelineExecGroupImpl(DB::PipelineExecutorContext&, DB::PipelineExecGroupBuilder&, DB::Context&, unsigned long) [tiflash+144990613]
                \tdbms/src/Flash/Planner/Plans/PhysicalJoinProbe.cpp:50
       0x89d55ce\tDB::PhysicalPlanNode::buildPipelineExecGroup(DB::PipelineExecutorContext&, DB::PipelineExecGroupBuilder&, DB::Context&, unsigned long) [tiflash+144528846]                \tdbms/src/Flash/Planner/PhysicalPlanNode.cpp:130 
       0x8991f7f\tDB::Pipeline::buildExecGroup(DB::PipelineExecutorContext&, DB::Context&, unsigned long) [tiflash+144252799]
                \tdbms/src/Flash/Pipeline/Pipeline.cpp:202
       0x89aca02\tDB::PlainPipelineEvent::scheduleImpl() [tiflash+144361986]
                \tdbms/src/Flash/Pipeline/Schedule/Events/PlainPipelineEvent.cpp:24
       0x89a5bc2\tDB::Event::schedule() [tiflash+144333762]
                \tdbms/src/Flash/Pipeline/Schedule/Events/Event.cpp:132
       0x89a6c5c\tDB::Event::finish() [tiflash+144338012]
                \tdbms/src/Flash/Pipeline/Schedule/Events/Event.cpp:193
       0x89b5780\tDB::EventTask::finalizeImpl() [tiflash+144398208]
                \tdbms/src/Flash/Pipeline/Schedule/Tasks/EventTask.cpp:41
       0x89bb8b2\tDB::Task::finalize() [tiflash+144423090]
                \tdbms/src/Flash/Pipeline/Schedule/Tasks/Task.cpp:162
       0x1eb30bb\tDB::TaskThreadPool<DB::CPUImpl>::loop(unsigned long) [tiflash+32190651]
                \tdbms/src/Flash/Pipeline/Schedule/ThreadPool/TaskThreadPool.cpp:61
       0x1eb3806\tvoid* std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, void (DB::TaskThreadPool<DB::CPUImpl>::*)(unsigned long), DB::TaskThreadPool<DB::CPUImpl>*, unsigned long> >(void*) [tiflash+32192518]
                \t/usr/local/bin/../include/c++/v1/thread:291
  0x7f352e8d617a\tstart_thread [libpthread.so.0+33146]"] [source=BaseDaemon] [thread_id=1296]

--- a/dbms/src/Flash/tests/gtest_join_executor.cpp
+++ b/dbms/src/Flash/tests/gtest_join_executor.cpp
@@ -1802,6 +1802,25 @@ try
 }
 CATCH
 
+TEST_F(JoinExecutorTestRunner, LeftJoinAggWithOtherCondition)
+try
+{
+    auto request_column
+        = context.scan("test_db", "l_table")
+              .join(
+                  context.scan("test_db", "r_table"),
+                  tipb::JoinType::TypeLeftOuterJoin,
+                  {col("join_c")},
+                  {},
+                  {},
+                  {And(lt(col("l_table.s"), col("r_table.s")), eq(col("l_table.join_c"), col("r_table.join_c")))},
+                  {})
+              .aggregation({Count(lit(static_cast<UInt64>(1)))}, {})
+              .build(context);
+    ASSERT_COLUMNS_EQ_UR(genScalarCountResults(2), executeStreams(request_column, 2));
+}
+CATCH
+
 TEST_F(JoinExecutorTestRunner, JoinWithExchangeReceiver)
 try
 {

--- a/dbms/src/Interpreters/Join.cpp
+++ b/dbms/src/Interpreters/Join.cpp
@@ -1029,16 +1029,19 @@ void Join::handleOtherConditions(
         for (size_t right_table_column : right_table_columns)
         {
             auto & column = block.getByPosition(right_table_column);
-            auto full_column
-                = column.column->isColumnConst() ? column.column->convertToFullColumnIfConst() : column.column;
-            if (!full_column->isColumnNullable())
+            if (output_column_names_set_after_finalize.contains(column.name))
             {
-                throw Exception("Should not reach here, the right table column for left join must be nullable");
+                auto full_column
+                    = column.column->isColumnConst() ? column.column->convertToFullColumnIfConst() : column.column;
+                if (!full_column->isColumnNullable())
+                {
+                    throw Exception("Should not reach here, the right table column for left join must be nullable");
+                }
+                auto current_column = full_column;
+                auto result_column = (*std::move(current_column)).mutate();
+                static_cast<ColumnNullable &>(*result_column).applyNegatedNullMap(*filter_column);
+                column.column = std::move(result_column);
             }
-            auto current_column = full_column;
-            auto result_column = (*std::move(current_column)).mutate();
-            static_cast<ColumnNullable &>(*result_column).applyNegatedNullMap(*filter_column);
-            column.column = std::move(result_column);
         }
         erase_useless_column(block);
         for (size_t i = 0; i < block.columns(); ++i)
@@ -1155,16 +1158,19 @@ void Join::handleOtherConditionsForOneProbeRow(Block & block, ProbeProcessInfo &
             for (size_t right_table_column : probe_process_info.cross_join_data->right_column_index_in_result_block)
             {
                 auto & column = block.getByPosition(right_table_column);
-                auto full_column
-                    = column.column->isColumnConst() ? column.column->convertToFullColumnIfConst() : column.column;
-                if (!full_column->isColumnNullable())
+                if (output_column_names_set_after_finalize.contains(column.name))
                 {
-                    throw Exception("Should not reach here, the right table column for left join must be nullable");
+                    auto full_column
+                        = column.column->isColumnConst() ? column.column->convertToFullColumnIfConst() : column.column;
+                    if (!full_column->isColumnNullable())
+                    {
+                        throw Exception("Should not reach here, the right table column for left join must be nullable");
+                    }
+                    auto current_column = full_column;
+                    auto result_column = (*std::move(current_column)).mutate();
+                    static_cast<ColumnNullable &>(*result_column).applyNegatedNullMap(*filter_column);
+                    column.column = std::move(result_column);
                 }
-                auto current_column = full_column;
-                auto result_column = (*std::move(current_column)).mutate();
-                static_cast<ColumnNullable &>(*result_column).applyNegatedNullMap(*filter_column);
-                column.column = std::move(result_column);
             }
             erase_useless_column(block);
         }
@@ -2641,17 +2647,16 @@ void Join::finalize(const Names & parent_require)
         updated_require.push_back(non_equal_conditions.other_eq_cond_from_in_name);
     if (!non_equal_conditions.other_cond_name.empty())
         updated_require.push_back(non_equal_conditions.other_cond_name);
-    auto keep_used_input_columns
-        = !isCrossJoin(kind) && (isNullAwareSemiFamily(kind) || isSemiFamily(kind) || isLeftOuterSemiFamily(kind));
     /// nullaware/semi join will reuse the input columns so need to let finalize keep the input columns
     if (non_equal_conditions.null_aware_eq_cond_expr != nullptr)
     {
-        non_equal_conditions.null_aware_eq_cond_expr->finalize(updated_require, keep_used_input_columns);
+        /// todo don't keep input columns for non-semi/non-nullaware join
+        non_equal_conditions.null_aware_eq_cond_expr->finalize(updated_require, true);
         updated_require = non_equal_conditions.null_aware_eq_cond_expr->getRequiredColumns();
     }
     if (non_equal_conditions.other_cond_expr != nullptr)
     {
-        non_equal_conditions.other_cond_expr->finalize(updated_require, keep_used_input_columns);
+        non_equal_conditions.other_cond_expr->finalize(updated_require, true);
         updated_require = non_equal_conditions.other_cond_expr->getRequiredColumns();
     }
     /// remove duplicated column

--- a/dbms/src/Interpreters/Join.cpp
+++ b/dbms/src/Interpreters/Join.cpp
@@ -2650,12 +2650,12 @@ void Join::finalize(const Names & parent_require)
     /// nullaware/semi join will reuse the input columns so need to let finalize keep the input columns
     if (non_equal_conditions.null_aware_eq_cond_expr != nullptr)
     {
-        /// todo don't keep input columns for non-semi/non-nullaware join
         non_equal_conditions.null_aware_eq_cond_expr->finalize(updated_require, true);
         updated_require = non_equal_conditions.null_aware_eq_cond_expr->getRequiredColumns();
     }
     if (non_equal_conditions.other_cond_expr != nullptr)
     {
+        /// todo don't keep input columns for non-semi/non-nullaware join
         non_equal_conditions.other_cond_expr->finalize(updated_require, true);
         updated_require = non_equal_conditions.other_cond_expr->getRequiredColumns();
     }

